waiter/src/waiter/interstitial.clj

```diff
@@ -0,0 +1,237 @@
+         (->> service-ids
+              (filter #(realized? (service-id->interstitial-promise %)))
+              (apply dissoc service-id->interstitial-promise)
+              (assoc interstitial-state :service-id->interstitial-promise)))
+       (swap! interstitial-state-atom))
+  (->> @interstitial-state-atom
+       :service-id->interstitial-promise
+       keys
+       set
+       (set/difference service-ids)))
+
+(defn install-interstitial-timeout!
+  "Resolves the interstitial-promise after a delay on interstitial-secs seconds."
+  [service-id interstitial-secs interstitial-promise]
+  (if (pos? interstitial-secs)
+    (async/go
+      (-> interstitial-secs t/seconds t/in-millis async/timeout async/<!)
+      (resolve-promise! service-id interstitial-promise :interstitial-timeout))
+    (log/error service-id "has opted out of interstitial, not installing timeout")))
+
+(defn- process-scheduler-messages
+  "Processes messages from the scheduler.
+   In particular, it resolves all promises for services which have healthy instances.
+   It also removes from the interstitial state any resolved promises for services which are no longer available."
+  [interstitial-state-atom service-id->service-description current-available-service-ids scheduler-messages]
+  (let [available-service-ids (->> scheduler-messages
+                                   (some (fn [[message-type message-data]]
+                                           (when (= :update-available-apps message-type)
+                                             (:available-apps message-data))))
+                                   set)
+        healthy-instance-service-ids (->> scheduler-messages
+                                          (map (fn [[message-type message-data]]
+                                                 (when (and (= :update-app-instances message-type)
+                                                            (seq (:healthy-instances message-data)))
+                                                   (:service-id message-data))))
+                                          (remove nil?))
+        service-ids-to-remove (set/difference current-available-service-ids available-service-ids)
+        removed-service-ids (remove-resolved-interstitial-promises! interstitial-state-atom service-ids-to-remove)]
+    (doseq [service-id available-service-ids]
+      (let [{:strs [interstitial-secs]} (service-id->service-description service-id)]
+        (when (pos? interstitial-secs)
+          (->> (partial install-interstitial-timeout! service-id interstitial-secs)
+               (ensure-service-interstitial! interstitial-state-atom service-id)))))
+    (doseq [service-id healthy-instance-service-ids]
+      (when-let [interstitial-promise (service-id->interstitial-promise @interstitial-state-atom service-id)]
+        (resolve-promise! service-id interstitial-promise :healthy-instance-found)))
+    (-> current-available-service-ids
```

Comment:
Is the value of this thread ever different than the value of `available-service-ids`?